compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -566,36 +576,37 @@ private BType getIntegerLiteralType(BLangLiteral literalExpr, Object literalValu
                     symTable.byteType);
             if (finiteTypeMatchingByte != symTable.semanticError) {
                 finiteType = finiteTypeMatchingByte;
-                BType setType = setLiteralValueAndGetType(literalExpr, finiteType);
+                BType setType = setLiteralValueAndGetType(literalExpr, finiteType, data);
                 if (literalExpr.isFiniteContext) {
                     // i.e., a match was found for a finite type
                     return setType;
                 }
             }
 
             Set<BType> memberTypes = ((BUnionType) expectedType).getMemberTypes();
-            return getTypeMatchingFloatOrDecimal(finiteType, memberTypes, literalExpr, (BUnionType) expectedType);
+            return getTypeMatchingFloatOrDecimal(finiteType, memberTypes, literalExpr, (BUnionType) expectedType, data);
         }
         return symTable.intType;
     }
 
-    private BType getTypeOfLiteralWithFloatDiscriminator(BLangLiteral literalExpr, Object literalValue, BType expType) {
+    private BType getTypeOfLiteralWithFloatDiscriminator(BLangLiteral literalExpr, Object literalValue,
+                                                         BType expType, AnalyzerData data) {
         String numericLiteral = NumericLiteralSupport.stripDiscriminator(String.valueOf(literalValue));
         if (!types.validateFloatLiteral(literalExpr.pos, numericLiteral)) {
-            resultType = symTable.semanticError;
-            return resultType;
+            data.resultType = symTable.semanticError;
+            return data.resultType;
```

Comment:
Let's return `symTable.semanticError`.

Also can you check whether we need to assign the type to `data.resultType`?
In this method this is the only place where we set the type to data, in other places we have just returned the type.